src/transformers/cache_utils.py

```diff
@@ -0,0 +1,89 @@
+from abc import ABC, abstractmethod
+from typing import Dict, List, Tuple, TypeVar
+import torch
+
+T = TypeVar("T")
+
+
+class Cache(ABC):
+    def __init__(self) -> None:
+        self.cache: Dict[int, Tuple[torch.Tensor]] = {}
+        self.layer_idx = 0
+
+    @abstractmethod
+    def update_pre_rotation(self, key_states: torch.Tensor, value_states: torch.Tensor) -> None:
```

Comment:
nit: "rotation" in the name implies the usage with Rotary Position Embeddings. I'd pick a more generic name :)

Reply:
It's actually related to that - this method exists to be called before position embeddings are rotated for RoPE and ALiBi approaches. This is a requirement for attention sinks, where the keys need to be cached before being rotated.

The name still isn't great though, and it probably shouldn't be an abstract method as it would likely only be used by some `AttentionSinkCache`.

Reply:
Is there no way to only have a `update(...)` function instead of forcing each cache to have to `update(...)` functions? Don't think that's very clean. Models like Llama/Falcon only support either RoPE or Alibi, can't we just call `update` here at the right position with a comment that it has to be done pre-rotation?

Reply:
Hmm I see, are there any other designs we can think of? E.g. could we just store the rotated k/v cache for the attention sink cache and then for the next step just back-rotate and forward-rotate it? Wouldn't back-rotate and forward-rotate it be the same as to just take the difference in angle for the next step and have the same result (just brainstorming, haven't looked in-detail here)

Reply:
I will be working on this - I think it's a good suggestion to help with simplicity, but I'm not a big fan of rotating 2 or 3 times rather than once.

Reply:
Successfully implemented this in ba87d7383a7a4fde292d073cff33133d659f61af.
I ended up doing some math to determine the cos/sin required to back-rotate and forward-rotate in one go, resulting in a solution with only a small performance decrease. However, moving this functionality out of `...Attention.forward` and into the `SinkCache` does mean that I have to investigate how easily this translates to other architectures.